### ISSUE TITLE
[Debian Jessie] kobo touch 905C: can not build

### ISSUE THREAD

**Issue by zaoqi-unsafe** [koreader/koreader#4398]
* KOReader version:github
* Device: kobo touch 905c with debian jessie

#### Steps to reproduce
```
marek@kobo:~/marek-zaoqi-init/koreader$ ./kodev build
v2018.12-11-gf020b12
grep: /proc/1/cgroup: 没有那个文件或目录
make -C base
make[1]: Entering directory '/home/marek/marek-zaoqi-init/koreader/base'
************ Building for MACHINE: "arm-linux-gnueabi-debug" **********
************ PATH: "/home/marek/scripts:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games" **********
************ CHOST: "" **********
install -d thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug
cd thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug && \
	cmake  \
	-DCC="/usr/bin/ccache gcc" -DCFLAGS="-Og -g -pipe  -march=native -fPIC" -DCXX="/usr/bin/ccache g++" -DCXXFLAGS="-Og -g -pipe  -march=native -fPIC -O3" \
	-DAR="ar" -DRANLIB="ranlib" -DSTDCPPLIB="/usr/lib/gcc/arm-linux-gnueabi/4.9/libstdc++.a" -DLDFLAGS="-Wl,-O1 -Wl,--as-needed -static-libstdc++" \
	-DZLIB_DIR=/home/marek/marek-zaoqi-init/koreader/base/thirdparty/zlib/build/arm-linux-gnueabi-debug/zlib-prefix/src/zlib -DZLIB=/home/marek/marek-zaoqi-init/koreader/base/build/arm-linux-gnueabi-debug/libs/libz.so.1 -DPNG_DIR=/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libpng/build/arm-linux-gnueabi-debug/libpng-prefix/src/libpng-build \
	-DLEPTONICA_DIR=/home/marek/marek-zaoqi-init/koreader/base/thirdparty/leptonica/build/arm-linux-gnueabi-debug/leptonica-prefix/src/leptonica -DTESSERACT_DIR=/home/marek/marek-zaoqi-init/koreader/base/thirdparty/tesseract/build/arm-linux-gnueabi-debug/tesseract-prefix/src/tesseract \
	/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt && \
	make
-- Configuring done
-- Generating done
-- Build files have been written to: /home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug
make[2]: Entering directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug'
make[3]: Entering directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug'
make[4]: Entering directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug'
make[4]: Leaving directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug'
make[4]: Entering directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug'
[ 12%] Performing build step for 'libk2pdfopt'
make[5]: Entering directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug/libk2pdfopt-prefix/src/libk2pdfopt'
sh: 0: Can't open ./autobuild
sh: 0: Can't open ./configure
Makefile:83: recipe for target 'liblept.so' failed
make[5]: *** [liblept.so] Error 127
make[5]: Leaving directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug/libk2pdfopt-prefix/src/libk2pdfopt'
CMakeFiles/libk2pdfopt.dir/build.make:104: recipe for target 'libk2pdfopt-prefix/src/libk2pdfopt-stamp/libk2pdfopt-build' failed
make[4]: *** [libk2pdfopt-prefix/src/libk2pdfopt-stamp/libk2pdfopt-build] Error 2
make[4]: Leaving directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug'
CMakeFiles/Makefile2:60: recipe for target 'CMakeFiles/libk2pdfopt.dir/all' failed
make[3]: *** [CMakeFiles/libk2pdfopt.dir/all] Error 2
make[3]: Leaving directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug'
Makefile:76: recipe for target 'all' failed
make[2]: *** [all] Error 2
make[2]: Leaving directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/libk2pdfopt/build/arm-linux-gnueabi-debug'
Makefile.third:212: recipe for target 'build/arm-linux-gnueabi-debug/libs/libk2pdfopt.so.2' failed
make[1]: *** [build/arm-linux-gnueabi-debug/libs/libk2pdfopt.so.2] Error 2
make[1]: Leaving directory '/home/marek/marek-zaoqi-init/koreader/base'
Makefile:62: recipe for target 'all' failed
make: *** [all] Error 2
Failed to build emulator! Try run with -v for more information.
marek@kobo:~/marek-zaoqi-init/koreader$ 
```
##### `crash.log` (if applicable)
`crash.log` is a file that is automatically created when KOReader crashes. It can
normally be found in the KOReader directory:

* `koreader/` directory for Kindle
* `.adds/koreader/` directory for Kobo
* `applications/koreader/` directory for Pocketbook

Please try to include the relevant sections in your issue description.
You can upload the whole `crash.log` file on GitHub by dragging and
dropping it onto this textbox, or you can use one of these paste services.
Please don't forget to set a high expiration date if you do.
https://pastebin.com https://slexy.org https://paste2.org https://fpaste.org
http://ix.io https://paste.kde.org https://paste.debian.net.

**Comment by zaoqi-unsafe** [comment]
> You didn't say which distro you're using? :-)
> 
> I think that's probably indicative of not having autoconf installed, see [here](https://github.com/koreader/koreader#building-prerequisites) for the building requirements.
> 
> ```
> sudo apt-get install build-essential git patch wget unzip \
> gettext autoconf automake cmake libtool nasm luarocks \
> libssl-dev libffi-dev libsdl2-dev libc6-dev-i386 xutils-dev linux-libc-dev:i386 zlib1g:i386
> ```

I have done it.
It is a kobo touch 905c running Debian jessie GNU/Linux armel with X11.

**Comment by pazos** [comment]
> It is a kobo touch 905c running Debian jessie GNU/Linux armel with X11.

wow, so you are building the emulator on an armel host. Keep that in mind. You won't get the fullscreen application, neither the suspend/poweroff/reboot functionality and you'll need to rely on 3rd party software to handle wireless connection.

**Comment by Frenzie** [comment]
Oh, you're building *on* the device? That's interesting! :-D

I'm not sure if there's any particular reason that shouldn't work, but I'd probably advise against it. It could easily take hours.

**Comment by Frenzie** [comment]
@pazos 
> Jessie worked fine to build the emulator, at least in x86_64.

I can't see a reason it shouldn't. :-) I personally may not have built KOReader on Jessie in well over a year,[1] but aside from some tiny fixes to make it build on Ubuntu 18.04+ *as well* nothing of significance has changed. Besides, we still test the build on the older Ubuntu 14.04 in the CI.

[1] I switched to Stretch shortly after it was released and I've been on Buster for quite some time now.

**Comment by pazos** [comment]
@Frenzie thanks for the info. I have a few debian/ubuntu VMs for testing stuff (I'm using Slackware-current because I'm slack). Didn't test recently but did a few builds on wheezy over the last two years.

I think that @zaoqi-unsafe will find issues trying to build on non intel hardware. To build the emulator we're using [`-march=native`](https://github.com/koreader/koreader-base/blob/1fa2b5715e2be7432e473940cc623d396aa9f5f7/Makefile.defs#L338) which seems not well supported on arm until GCC8 -> https://www.phoronix.com/scan.php?page=news_item&px=GCC-8-march-native-ARM.

**Comment by Frenzie** [comment]
Shouldn't that just mean you get an unoptimized build?

**Comment by zaoqi-unsafe** [comment]
I deleted it and did retry.
Then I got
```
marek@kobo:~/marek-zaoqi-init/koreader$ ./kodev build
v2018.12-12-g8fec0f3
grep: /proc/1/cgroup: 没有那个文件或目录
make -C base
make[1]: Entering directory '/home/marek/marek-zaoqi-init/koreader/base'
************ Building for MACHINE: "arm-linux-gnueabi-debug" **********
************ PATH: "/home/marek/scripts:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games" **********
************ CHOST: "" **********
install -d thirdparty/freetype2/build/arm-linux-gnueabi-debug
cd thirdparty/freetype2/build/arm-linux-gnueabi-debug && \
	cmake -DCC="/usr/bin/ccache gcc" -DCXX="/usr/bin/ccache g++" -DCFLAGS="-Og -g -pipe  -march=native -fPIC"\
	-DCXXFLAGS="-Og -g -pipe  -march=native -fPIC" -DLDFLAGS="-Wl,-O1 -Wl,--as-needed -static-libstdc++" -DCHOST= \
	/home/marek/marek-zaoqi-init/koreader/base/thirdparty/freetype2 && \
	make
-- Configuring done
-- Generating done
-- Build files have been written to: /home/marek/marek-zaoqi-init/koreader/base/thirdparty/freetype2/build/arm-linux-gnueabi-debug
make[2]: Entering directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/freetype2/build/arm-linux-gnueabi-debug'
make[3]: Entering directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/freetype2/build/arm-linux-gnueabi-debug'
make[4]: Entering directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/freetype2/build/arm-linux-gnueabi-debug'
make[4]: Leaving directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/freetype2/build/arm-linux-gnueabi-debug'
make[4]: Entering directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/freetype2/build/arm-linux-gnueabi-debug'
[ 12%] Performing patch step for 'freetype2'
ERROR: Your version of the `autoconf' tool is too old.
       Minimum version 2.62 is required (yours is version 2.13).
       Please upgrade or use the AUTOCONF variable to point to a more recent one.

CMakeFiles/freetype2.dir/build.make:93: recipe for target 'freetype2-prefix/src/freetype2-stamp/freetype2-patch' failed
make[4]: *** [freetype2-prefix/src/freetype2-stamp/freetype2-patch] Error 1
make[4]: Leaving directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/freetype2/build/arm-linux-gnueabi-debug'
CMakeFiles/Makefile2:60: recipe for target 'CMakeFiles/freetype2.dir/all' failed
make[3]: *** [CMakeFiles/freetype2.dir/all] Error 2
make[3]: Leaving directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/freetype2/build/arm-linux-gnueabi-debug'
Makefile:76: recipe for target 'all' failed
make[2]: *** [all] Error 2
make[2]: Leaving directory '/home/marek/marek-zaoqi-init/koreader/base/thirdparty/freetype2/build/arm-linux-gnueabi-debug'
Makefile.third:17: recipe for target 'build/arm-linux-gnueabi-debug/libs/libfreetype.so.6' failed
make[1]: *** [build/arm-linux-gnueabi-debug/libs/libfreetype.so.6] Error 2
make[1]: Leaving directory '/home/marek/marek-zaoqi-init/koreader/base'
Makefile:62: recipe for target 'all' failed
make: *** [all] Error 2
Failed to build emulator! Try run with -v for more information.
```
After updating, nothing have changed.

**Comment by zaoqi-unsafe** [comment]
```
sudo dpkg -i autoconf_2.69-10~bpo8+1_all.deb 
```

**Comment by Frenzie** [comment]
That's very odd since Debian Jessie comes with 2.69? As stated though, if it works it'll be a tad suboptimal since you'd be missing the device optimizations.

(And just as I was about to post this you confirm it's 2.69.)

**Comment by zaoqi-unsafe** [comment]
@Frenzie My intel/AMD compuers were broken.

**Comment by zaoqi-unsafe** [comment]
The device optimizations can kill debian.It is not a tad suboptimal

**Comment by zaoqi-unsafe** [comment]
```
marek@kobo:~/marek-zaoqi-init/koreader$ autoconf --version

Autoconf version 2.13
---
Autoconf 2.13 chosen by Debian wrapper script.
For information and tuning advice see autoconf(1).
```
!

**Comment by Frenzie** [comment]
Ah yes, I see the same is still true on Debian Buster, although I'm not sure why your Jessie would default to the oldest option:

```bash
$ apt search autoconf
autoconf/testing,testing,unstable,unstable,now 2.69-11 all [installed]
  automatic configure script builder

[…]

autoconf2.13/testing,testing,unstable,unstable 2.13-68 all
  automatic configure script builder (obsolete version)

autoconf2.64/testing,testing,unstable,unstable 2.64+dfsg-1 all
  automatic configure script builder (obsolete version)
```
I'd just remove the ancient autoconf (what's it good for?), but you could also look into the regular `update-alternatives --config autoconf` route.

> The device optimizations can kill debian.It is not a tad suboptimal

Well no, that was the diplomatic way of phrasing it. I still phrased that badly or at least incompletely though. *Debian* should probably be fine. It's the X11/Xorg part that'll make the device experience much worse, assuming the program will run at all.

On a regular desktop you might be able to go fullscreen by pressing Alt+F11, but on a device that'd probably require hardware that supports USB OTG combined with a kernel that does. It also wouldn't work with controlling wifi and various other things. But *if* it works, I suppose it should still function reasonably well as a touch-friendly, E Ink friendly reader. Also, however, without certain E Ink optimizations.

The program should still be more suitable for reading than many other alternatives within an X11 environment. Anyway, I'd certainly be interested to hear the results. I just wouldn't get my hopes up too much. :-)

**Comment by zaoqi** [comment]
@Frenzie 
In my awesome 3.4, I can press the physical button to go fullscreen.

**Comment by Frenzie** [comment]
Every command in kodev has a `--help` flag. In this case you'll want `--no-debug`.

**Comment by zaoqi** [comment]
It runs, but looks strange.

**Comment by zaoqi** [comment]
"No reader engine for this file or invalid file"
"Q" "W" ...
I can not select a file.
"Set ... as HOME DIC..." I can not touch "Set as HOME"

**Comment by Frenzie** [comment]
In that case there was a problem with crengine, meaning the quickstart guide couldn't open. I'm not really sure why that would be or why the build would (seemingly) complete in that case.

**Comment by zaoqi** [comment]
Could you  build it for me?

**Comment by Frenzie** [comment]
Just grab the Kobo build? It should run fine using SDL, just ignore the installation instructions.

**Comment by Frenzie** [comment]
It doesn't make much sense that it would find SDL one way and not the other. You should run it with --debug so you can see the full output.

But since you didn't want to compile crengine again, did you replacing the lib with the one from the Kobo build?

**Comment by zaoqi-unsafe** [comment]
>           It doesn't make much sense that it would find SDL one way and not the other. You should run it with --debug so you can see the full output.
> But since you didn't want to compile crengine again, did you replacing the lib with the one from the Kobo build?
The kobo firmware is armel, and can not be upgraded easily because of that debian image.

**Comment by Frenzie** [comment]
None of that has anything to do with the firmware. I apologize if I was unclear.

When you built the KOReader "emulator", it would have put the libraries in `koreader-emulator-arm-something-linux-gnu/koreader/libs/`. Okay, that's actually a symlink to `base/build/arm-etc` but that doesn't really matter. The list of libraries in there should look something like this, but you can also see that for yourself in the precompiled Kobo package:

```
~/src/kobo/koreader/base/build/x86_64-linux-gnu-debug/libs$ ls
libblitbuffer.so    libgif.so.7          liblept.so.4       libturbojpeg.so
libcrengine.so      libharfbuzz.so.0     liblodepng.so      libwrap-mupdf.so
libcrypto.so.1.0.0  libjpeg.so.8         libmupdf.so        libzmq.so.4
libczmq.so.1        libk2pdfopt.so.2     libpng16.so.16     libz.so.1
libdjvulibre.so.21  libkoreader-cre.so   libsqlite3.so      libzyre.so.1
libfmq.so.1         libkoreader-djvu.so  libssl.so.1.0.0
libfreetype.so.6    libkoreader-lfs.so   libtesseract.so.3
```

Now it's unclear to me what could've gone wrong in your build, like why it would've completed at all if there was a problem.

Anyway, if one of the libs is absent or broken, you could try building it again, which might work if it failed due to being out of memory or some such. But chances are it would also work when mixing libraries with the ones from the prebuilt Kobo binary package.

But if the emulator is (mostly) working, I don't understand why the precompiled Kobo binary would not. Are you actually running it in the same environment?

**Comment by pazos** [comment]
> Anyway, if one of the libs is absent or broken, you could try building it again, which might work if it failed due to being out of memory or some such. But chances are it would also work when mixing libraries with the ones from the prebuilt Kobo binary package.

Kobo is armhf, so it won't work on armel (softfp). But kindle binaries should work if the proccesor supports SIMD extensions or cervantes binaries if not.

**Comment by Frenzie** [comment]
Oh, I'd glanced over that. Why would you use armel?

**Comment by pazos** [comment]
@zaoqi @zaoqi-unsafe You can try the package at https://mega.nz/#!XyQAnaiQ and please provide feedback

to install it use `sudo dpkg -i path/to/koreader-*.deb`. If the command fails it is because you don't have sdl installed. Fix it with `sudo apt-get -f install` to fetch package dependencies and finish the installation.

Once the program is installed you can use the command `koreader` from a terminal or open its desktop file (placed in /usr/share/applications/koreader.desktop)

**Comment by zaoqi-unsafe** [comment]
@pazos
These website is blocked by the government.
So I download the debian image from pan.baidu.com and I can not download That file.
If I can download, I will test it on kobo touch 905C,kobo aura hd and KT3 running Ubuntu/Debian

**Comment by Frenzie** [comment]
This is the file. Just remove the .zip at the end.

[koreader-2018.12-38-armel.deb.zip](https://github.com/koreader/koreader/files/2727424/koreader-2018.12-38-armel.deb.zip)

**Comment by zaoqi-unsafe** [comment]
> This is the file. Just remove the .zip at the end.
> 
> [koreader-2018.12-38-armel.deb.zip](https://github.com/koreader/koreader/files/2727424/koreader-2018.12-38-armel.deb.zip)

Kindle and aura hd is armhf

**Comment by NiLuJe** [comment]
Sidebar: Kindle is *NOT* hf.

**Comment by NiLuJe** [comment]
Hmm, not necessarily, as long as you don't mix fp ABIs IIRC.

**Comment by Frenzie** [comment]
@NiLuJe 

> Sidebar: Kindle is NOT hf (... at least as far as their stock FW are concerned ;)).

Oh, I thought the CPU was? Anyway, my point was that it's the Debian build on this Kobo Touch that's using a different arch than our regular builds and whether it's stuck in a regular zip file or a Debian package (which is also a zip file) probably doesn't really matter too much for some quick experimentation. Anyway, I told my computer to build a Debian armhf release package while I go wash the dishes and stuff, so with any luck it'll be done by then.

**Comment by Frenzie** [comment]
Huh, that's weird. Now GH isn't letting me upload a file larger than 10 MB…

**Comment by zaoqi-unsafe** [comment]
```
marek@kobo:~/downloads$ mv koreader-2018.12-38-armel.deb.zip koreader-2018.12-38-armel.deb
marek@kobo:~/downloads$ file koreader-2018.12-38-armel.deb 
koreader-2018.12-38-armel.deb: Debian binary package (format 2.0)
marek@kobo:~/downloads$ sudo dpkg -i koreader-2018.12-38-armel.deb 
正在选中未选择的软件包 koreader。
(正在读取数据库 ... 系统当前共安装有 92930 个文件和目录。)
正准备解包 koreader-2018.12-38-armel.deb  ...
正在解包 koreader (2018.12-38) ...
dpkg: 依赖关系问题使得 koreader 的配置工作不能继续：
 koreader 依赖于 libsdl2-2.0-0；然而：
  未安装软件包 libsdl2-2.0-0。

dpkg: 处理软件包 koreader (--install)时出错：
 依赖关系问题 - 仍未被配置
正在处理用于 desktop-file-utils (0.22-1) 的触发器 ...
正在处理用于 mime-support (3.58) 的触发器 ...
在处理时有错误发生：
 koreader
marek@kobo:~/downloads$ sudo apt install -f
正在读取软件包列表... 完成
正在分析软件包的依赖关系树       
正在读取状态信息... 完成       
正在更正依赖关系... 完成
将会安装下列额外的软件包：
  libsdl2-2.0-0
下列【新】软件包将被安装：
  libsdl2-2.0-0
升级了 0 个软件包，新安装了 1 个软件包，要卸载 0 个软件包，有 0 个软件包未被升级。
有 1 个软件包没有被完全安装或卸载。
需要下载 301 kB 的软件包。
解压缩后会消耗掉 1,070 kB 的额外空间。
您希望继续执行吗？ [Y/n] 
错误 http://mirrors.ustc.edu.cn/debian/ jessie/main libsdl2-2.0-0 armel 2.0.2+dfsg1-6
  无法解析域名“mirrors.ustc.edu.cn”
E: 无法下载 http://mirrors.ustc.edu.cn/debian/pool/main/libs/libsdl2/libsdl2-2.0-0_2.0.2+dfsg1-6_armel.deb  无法解析域名“mirrors.ustc.edu.cn”

E: 有几个软件包无法下载，您可以运行 apt-get update 或者加上 --fix-missing 的选项再试试？
marek@kobo:~/downloads$ sudo apt install -f -y
正在读取软件包列表... 完成
正在分析软件包的依赖关系树       
正在读取状态信息... 完成       
正在更正依赖关系... 完成
将会安装下列额外的软件包：
  libsdl2-2.0-0
下列【新】软件包将被安装：
  libsdl2-2.0-0
升级了 0 个软件包，新安装了 1 个软件包，要卸载 0 个软件包，有 0 个软件包未被升级。
有 1 个软件包没有被完全安装或卸载。
需要下载 301 kB 的软件包。
解压缩后会消耗掉 1,070 kB 的额外空间。
获取：1 http://mirrors.ustc.edu.cn/debian/ jessie/main libsdl2-2.0-0 armel 2.0.2+dfsg1-6 [301 kB]
下载 301 kB，耗时 0秒 (335 kB/s)       
正在选中未选择的软件包 libsdl2-2.0-0:armel。
(正在读取数据库 ... 系统当前共安装有 93887 个文件和目录。)
正准备解包 .../libsdl2-2.0-0_2.0.2+dfsg1-6_armel.deb  ...
正在解包 libsdl2-2.0-0:armel (2.0.2+dfsg1-6) ...
正在设置 libsdl2-2.0-0:armel (2.0.2+dfsg1-6) ...
正在设置 koreader (2018.12-38) ...
正在处理用于 libc-bin (2.19-18+deb8u10) 的触发器 ...
marek@kobo:~/downloads$ ls
koreader-2018.12-38-armel.deb
koreader-kindle-arm-kindle5-linux-gnueabi-v2018.12.zip
尼尔·波兹曼：娱乐至死.mobi
尼尔·波兹曼：娱乐至死.pdf
marek@kobo:~/downloads$ koreader 
/usr/bin/koreader:行2: 警告:setlocale: LC_ALL: 无法改变区域选项 (en_US.UTF-8): 没有那个文件或目录
mkdir: created directory '/home/marek/.config/koreader'
mkdir: created directory '/home/marek/.config/koreader/fonts'
---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 01/06/19-11:00:57
 [*] Version: v2018.12-38-g9b197b13_2019-01-02

ffi.load: blitbuffer
ffi.load: SDL2
ffi.load (warning): libSDL2.so: cannot open shared object file: No such file or directory
ffi.load: SDL
ffi.load (warning): libSDL.so: cannot open shared object file: No such file or directory
./luajit: frontend/device.lua:45: Could not find hardware abstraction for this platform. If you are trying to run the emulator, please ensure SDL is installed.
stack traceback:
	[C]: in function 'error'
	frontend/device.lua:45: in function 'probeDevice'
	frontend/device.lua:48: in main chunk
	[C]: in function 'require'
	frontend/ui/widget/container/widgetcontainer.lua:19: in main chunk
	[C]: in function 'require'
	frontend/ui/widget/container/centercontainer.lua:5: in main chunk
	[C]: in function 'require'
	frontend/ui/widget/button.lua:19: in main chunk
	[C]: in function 'require'
	frontend/ui/widget/buttontable.lua:2: in main chunk
	[C]: in function 'require'
	frontend/ui/widget/confirmbox.lua:21: in main chunk
	[C]: in function 'require'
	./reader.lua:96: in main chunk
	[C]: at 0x00013bb9
marek@kobo:~/downloads$ 
```

**Comment by zaoqi-unsafe** [comment]
The buttons PgUp and PgDn seem to control it, but it is slow.

**Comment by zaoqi-unsafe** [comment]
I pressed a link multiple times and then it turned to the first page.

**Comment by pazos** [comment]
It works!!

You need to create a new sdl device with hasKeys=no and isTouchscreen=yes. Please look at frontend/device/sdl/device.lua.

Your device is recognized as a Linux device (with an standard keyboard and a pointer device = mouse). You need to change the device probe logic to check for your desired environmental variable before KO_MULTIUSER and return your new created device.

**Comment by zaoqi-unsafe** [comment]
kernel and .config/awesome/rc.lua are custom.Others are debian

**Comment by zaoqi-unsafe** [comment]
@pazos So I think that this device should be considered as a normal device running debian GNU/Linux jessie, not a special device.

**Comment by pazos** [comment]
> So I think that this device should be considered as a normal device running Linux, not a special device.

Then attach an usb keyboard/mouse and you're done

You're welcome to ask questions related to what you want to achieve but don't expect others to do the work for you.

**Comment by Frenzie** [comment]
> You can create a new sdl device with hasKeys=no

Doesn't really matter except visually for the quick access keys though.

It's weird that touch input isn't working. KOReader should work with both multitouch and click events in SDL. But whatever the problem, it's unlikely to be with KOReader imo.

Anyway, if you want to check up on what events SDL is receiving (if any), stick some debugging io.write/print commands around here:

https://github.com/koreader/koreader-base/blob/1b49222c7c5f292eb28dccaabff15d26c273b48e/ffi/SDL2_0.lua#L237-L239

They probably be one of these:

https://github.com/koreader/koreader-base/blob/1b49222c7c5f292eb28dccaabff15d26c273b48e/ffi/SDL2_0.lua#L274-L291

If there is a problem in KOReader then it shouldn't take very long to sort out that way. If you're only getting multitouch events into SDL, for example, I never worked on proper support for that in front but it might still be relatively easy to jerry-rig. Although afaik that's really just about multitouch gestures and for single finger up/down that's just what the driver should output.

https://hg.libsdl.org/SDL/file/default/docs/README-gesture.md

**Comment by zaoqi** [comment]
Koreader can't find the SDL of Debian .Maybe SDL should be placed in koreader.deb?

**Comment by Frenzie** [comment]
The link you made is just fine.

**Comment by zaoqi-unsafe** [comment]
@Frenzie

**Comment by Frenzie** [comment]
@zaoqi-unsafe  Sorry, what is the question precisely? One of the various existing binaries should likely work for basic testing.

**Comment by zaoqi-unsafe** [comment]
```
$ cat /etc/apt/sources.list
# See http://help.ubuntu.com/community/UpgradeNotes for how to upgrade to
# newer versions of the distribution.
deb http://ftp.cn.debian.org/ubuntu-ports/ bionic main restricted
deb-src http://ftp.cn.debian.org/ubuntu-ports/ bionic main restricted

## Major bug fix updates produced after the final release of the
## distribution.
deb http://ftp.cn.debian.org/ubuntu-ports/ bionic-updates main restricted
deb-src http://ftp.cn.debian.org/ubuntu-ports/ bionic-updates main restricted

## N.B. software from this repository is ENTIRELY UNSUPPORTED by the Ubuntu
## team. Also, please note that software in universe WILL NOT receive any
## review or updates from the Ubuntu security team.
deb http://ftp.cn.debian.org/ubuntu-ports/ bionic universe
deb-src http://ftp.cn.debian.org/ubuntu-ports/ bionic universe
deb http://ftp.cn.debian.org/ubuntu-ports/ bionic-updates universe
deb-src http://ftp.cn.debian.org/ubuntu-ports/ bionic-updates universe

## N.B. software from this repository is ENTIRELY UNSUPPORTED by the Ubuntu
## team, and may not be under a free licence. Please satisfy yourself as to
## your rights to use the software. Also, please note that software in
## multiverse WILL NOT receive any review or updates from the Ubuntu
## security team.
deb http://ftp.cn.debian.org/ubuntu-ports/ bionic multiverse
deb-src http://ftp.cn.debian.org/ubuntu-ports/ bionic multiverse
deb http://ftp.cn.debian.org/ubuntu-ports/ bionic-updates multiverse
deb-src http://ftp.cn.debian.org/ubuntu-ports/ bionic-updates multiverse

## N.B. software from this repository may not have been tested as
## extensively as that contained in the main release, although it includes
## newer versions of some applications which may provide useful features.
## Also, please note that software in backports WILL NOT receive any review
## or updates from the Ubuntu security team.
deb http://ftp.cn.debian.org/ubuntu-ports/ bionic-backports main restricted universe multiverse
deb-src http://ftp.cn.debian.org/ubuntu-ports/ bionic-backports main restricted universe multiverse

## Uncomment the following two lines to add software from Canonical's
## 'partner' repository.
## This software is not part of Ubuntu, but is offered by Canonical and the
## respective vendors as a service to Ubuntu users.
# deb http://archive.canonical.com/ubuntu bionic partner
# deb-src http://archive.canonical.com/ubuntu bionic partner

deb http://ftp.cn.debian.org/ubuntu-ports/ bionic-security main restricted
deb-src http://ftp.cn.debian.org/ubuntu-ports/ bionic-security main restricted
deb http://ftp.cn.debian.org/ubuntu-ports/ bionic-security universe
deb-src http://ftp.cn.debian.org/ubuntu-ports/ bionic-security universe
deb http://ftp.cn.debian.org/ubuntu-ports/ bionic-security multiverse
deb-src http://ftp.cn.debian.org/ubuntu-ports/ bionic-security multiverse
$ cat /etc/apt/sources.list.d/debian.list 
deb http://ftp.cn.debian.org/debian stretch main contrib non-free
deb-src http://ftp.cn.debian.org/debian stretch main contrib non-free

deb http://ftp.cn.debian.org/debian-security/ stretch/updates main contrib non-free
deb-src http://ftp.cn.debian.org/debian-security/ stretch/updates main contrib non-free

deb http://ftp.cn.debian.org/debian stretch-updates main contrib non-free
deb-src http://ftp.cn.debian.org/debian stretch-updates main contrib non-free
```
```
$ sudo dpkg --add-architecture armel
$ sudo apt update
$ sudo dpkg -i koreader-2018.12-38-armel.deb
$ sudo aptitude install -f
[sudo] zaoqi 的密码： 
下列仅部分安装的软件包将被配置： 
  koreader:armel{b} 
将不会安装，升级或者删除任何软件包。
0 个软件包被升级，新安装 0 个， 0 个将被删除， 同时 4 个将不升级。
需要获取 0 B 的存档。 解包后将要使用 0 B。
下列软件包存在未满足的依赖关系：
 koreader:armel : 依赖: libsdl2-2.0-0:armel but it is not going to be installed
打开：16；关闭；74；defer: 10; 冲突；10       .下列动作将解决这些依赖关系：

     删除 下列软件包：                  
1)     koreader:armel [2018.12-38 (now)]



是否接受该解决方案？[Y/n/q/?] n
打开：5030；关闭；12231；defer: 181; 冲突；23 O在限定的时间内没能找到解决方案。加把劲再试试？[Y/n] 
依赖关系分析中...
打开：8343；关闭；19388；defer: 181; 冲突；23 O^C
```
@Frenzie

**Comment by zaoqi-unsafe** [comment]
```
zaoqi@kindle:~$ sudo apt install libsdl2-2.0-0:armel
正在读取软件包列表... 完成
正在分析软件包的依赖关系树       
正在读取状态信息... 完成       
您也许需要运行“apt --fix-broken install”来修正上面的错误。
下列软件包有未满足的依赖关系：
 libsdl2-2.0-0:armel : 依赖: libasound2:armel (>= 1.0.16) 但是它将不会被安装
                       依赖: libc6:armel (>= 2.15) 但是它将不会被安装
                       依赖: libpulse0:armel (>= 0.99.1) 但是它将不会被安装
                       依赖: libsndio6.1:armel (>= 1.1.0) 但是它将不会被安装
                       依赖: libwayland-client0:armel (>= 1.9.91) 但是它将不会被安装
                       依赖: libwayland-cursor0:armel (>= 1.0.2) 但是它将不会被安装
                       依赖: libwayland-egl1-mesa:armel (>= 10.0.2) 但是它将不会被安装 或
                               libwayland-egl1:armel
                       依赖: libx11-6:armel (>= 2:1.2.99.901) 但是它将不会被安装
                       依赖: libxcursor1:armel (> 1.1.2) 但是它将不会被安装
                       依赖: libxext6:armel 但是它将不会被安装
                       依赖: libxi6:armel (>= 2:1.2.99.4) 但是它将不会被安装
                       依赖: libxinerama1:armel 但是它将不会被安装
                       依赖: libxkbcommon0:armel (>= 0.5.0) 但是它将不会被安装
                       依赖: libxrandr2:armel (>= 2:1.2.99.3) 但是它将不会被安装
                       依赖: libxss1:armel 但是它将不会被安装
                       依赖: libxxf86vm1:armel 但是它将不会被安装
E: 有未能满足的依赖关系。请尝试不指明软件包的名字来运行“apt --fix-broken install”(也可以指定一个解决办法)。
zaoqi@kindle:~$ sudo apt install -f
正在读取软件包列表... 完成
正在分析软件包的依赖关系树       
正在读取状态信息... 完成       
正在修复依赖关系... 完成
下列软件包将被【卸载】：
  koreader:armel
升级了 0 个软件包，新安装了 0 个软件包，要卸载 1 个软件包，有 4 个软件包未被升级。
有 1 个软件包没有被完全安装或卸载。
解压缩后将会空出 68.7 MB 的空间。
您希望继续执行吗？ [Y/n] 
(正在读取数据库 ... 系统当前共安装有 49285 个文件和目录。)
正在卸载 koreader:armel (2018.12-38) ...
正在处理用于 mime-support (3.60ubuntu1) 的触发器 ...
zaoqi@kindle:~$ 

```

**Comment by Frenzie** [comment]
What I mean is that you have precompiled binaries for various architectures in https://github.com/koreader/koreader/releases/tag/v2019.01

Whether they say "Kobo," "Kindle" or even "Android" doesn't matter that much for basic testing. My Android binaries work on Kobo (and PC, for x86), and vice versa.

Based on the fact that you ran `dpkg --add-architecture armel` I can only assume you meant to ask for an armhf Debian package? You really can't just say that you have a Kindle with Ubuntu. As @NiLuJe pointed out Kindle normally implies armel, which is already present in the various Kindle packages. In your case then, that would've logically implied trying the Kobo package 

Anyway, I shared the armhf deb [here](https://github.com/Frenzie/koreader/releases/tag/v2019.01.29).

**Comment by pazos** [comment]
I labeled this issue out of date since the title [Debian Jessie] kobo touch 905C: can not build, doesn't apply anymore. I didn't test it on the same device (a kobo touch) because I don't have one, but it shouldn't matter as long as the OS is debian and the prerequisites are installed. I doubt that anyone would want to build KOReader in a low powered arm device because it is slow (for the record: an optimized native built on a raspberry pi 3 will take more that 8 hours and will require a swapfile of *at least* a couple of GB). Since ereaders have low memory and it is not a good idea to create a swapfile on a sd card I would label this as invalid too)

In short: do not try to build KOReader on low-end arm devices. Just use a desktop/laptop and cross compile it.

**Comment by Frenzie** [comment]
You almost make me want to build it on my phone. Almost.

(The Zenfone 2 has a pretty speedy CPU and 4 GB of RAM. With something like Debian or Ubuntu in chroot it should theoretically go quite well, like a slow but not entirely outdated laptop.)

**Comment by zaoqi-unsafe** [comment]
@Frenzie I run Ubuntu using chroot
https://github.com/zaoqi-unsafe/kindle-ubuntu
~/.config/awesome https://github.com/zaoqi-unsafe/AwesomeTouch

**Comment by zaoqi-unsafe** [comment]
@Frenzie /mnt/us have been formatted into ext3

**Comment by zaoqi-unsafe** [comment]
```
zaoqi@kindle:~$ apt show libc6
Package: libc6
Version: 2.27-3ubuntu1
Priority: required
Section: libs
Source: glibc
Origin: Ubuntu
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: GNU Libc Maintainers <debian-glibc@lists.debian.org>
Bugs: https://bugs.launchpad.net/ubuntu/+filebug
Installed-Size: 8,747 kB
Provides: libc6-armhf
Depends: libgcc1
Suggests: glibc-doc, debconf | debconf-2.0, locales
Conflicts: openrc (<< 0.27-2~)
Breaks: hurd (<< 1:0.5.git20140203-1), libtirpc1 (<< 0.2.3), locales (<< 2.27), locales-all (<< 2.27), nscd (<< 2.27)
Homepage: https://www.gnu.org/software/libc/libc.html
Task: minimal
Supported: 5y
Download-Size: 2,207 kB
APT-Manual-Installed: no
APT-Sources: http://ftp.cn.debian.org/ubuntu-ports bionic/main armhf Packages
Description: GNU C 语言运行库：共享库
 本软件包包含了系统中几乎所有程序都会用到的标准运行库，包括标准 C 语言共享库和标准数学 库，以及许多其它的运行库。

N: 有 2 条附加记录。请加上 ‘-a’ 参数来查看它们
zaoqi@kindle:~$ koreader 
/usr/bin/koreader: 行 2: 警告: setlocale: LC_ALL: 无法改变区域选项 (en_US.UTF-8): 没有那个文件或目录
---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 02/02/19-11:04:57
./luajit: error loading module 'libs/libkoreader-lfs' from file './libs/libkoreader-lfs.so':
	/lib/arm-linux-gnueabihf/libc.so.6: version `GLIBC_2.28' not found (required by ./libs/libkoreader-lfs.so)
stack traceback:
	[C]: at 0x76fbb6b9
	[C]: in function 'require'
	./datastorage.lua:3: in main chunk
	[C]: in function 'require'
	./reader.lua:18: in main chunk
	[C]: at 0x76f80a65
zaoqi@kindle:~$ 
```

**Comment by pazos** [comment]
@zaoqi-unsafe: congratulations. I'm closing this issue since you managed to achieve want to want. If you have more related questions please comment here instead of open a new issue. 

Thanks!

**Comment by zaoqi** [comment]
@Frenzie *.deb should be rebuilt

**Comment by Frenzie** [comment]
Please note that "should" comes across as impolite in English.

If China doesn't block GitLab, all nightly builds are created using the GitLab infrastructure at <https://gitlab.com/koreader/nightly-builds/-/jobs>. Pick the one you want and download the artifacts.

![screenshot_2019-02-04_16-06-23](https://user-images.githubusercontent.com/202757/52216626-fad55080-2896-11e9-8479-672a8b0e7256.png)

Usually it's easier to go to <http://build.koreader.rocks/download/nightly/> instead, and for Android that's the only place where you'll find properly signed packages. But also, GitLab somehow manages to keep an archive that goes back at least a year, whereas our own server is barely even able to hold half a couple of weeks of builds.
